File: DSA - Data Structures And Algorithm/C/Binary Tree/Inorder Traversal on Binary Tree (Non recursive).c
// In this program we will perform inorder traversal on binary tree (non recursively) or we can say using stack.

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int val;
    struct Node* left;
    struct Node* right;
};

struct Node* newNode(int data) {
    struct Node* node = (struct Node*)malloc(sizeof(struct Node));
    node->val = data;
    node->left = NULL;
    node->right = NULL;
    return node;
}

struct Node* BuildTree() {
    int data;
    printf("Enter data: ");
    scanf("%d", &data);

    if (data == -1) return NULL;

    struct Node* root = newNode(data);

    printf("Enter data in left of %d\n", data);
    root->left = BuildTree();

    printf("Enter data in right of %d\n", data);
    root->right = BuildTree();

    return root;
}

void Inorder(struct Node* root) {
    struct Node* stack[100];
    int top = -1;
    struct Node* current = root;

    while (current != NULL || top != -1) {
        while (current != NULL) {
            stack[++top] = current; 
            current = current->left; 
        }

        current = stack[top--]; 
        printf("%d ", current->val); 

        current = current->right; 
    }
}

int main() {
    struct Node* root = NULL;
    root = BuildTree();
    printf("Inorder Traversal: ");
    Inorder(root);
    return 0;
}

File: DSA - Data Structures And Algorithm/Python/Binary Tree/Inorder Traversal on Binary Tree (Non recursive).py
# In this program we will perform inorder traversal on binary tree (non-recursively) using a stack

class Node:
    def __init__(self, data):
        self.val = data
        self.left = None
        self.right = None

def build_tree():
    data = int(input("Enter data: "))
    
    if data == -1:
        return None
    
    root = Node(data)
    
    print(f"Enter data in left of {data}")
    root.left = build_tree()
    
    print(f"Enter data in right of {data}")
    root.right = build_tree()
    
    return root

def inorder_non_recursive(root):
    stack = []
    current = root

    while current or stack:
        while current:
            stack.append(current)
            current = current.left
        
        current = stack.pop()
        print(current.val, end=" ") 
        
        current = current.right

root = build_tree()
print("Inorder Traversal (Non-Recursive):", end=" ")
inorder_non_recursive(root)
print()
